### ISSUE TITLE
Results.indexOf crashing app

### ISSUE THREAD

**Issue by kbduvall** [realm/realm-cocoa#3319]
I'm trying to get the index of the first occurrence of `Predicate` in a `Results` object using `indexOf(predicate: NSPredicate)` but it's crashing every time. 

Model Object:
```Swift
class Service: Object {
	dynamic var timestamp: NSDate?
	dynamic var server: Server?
	dynamic var backgroundMonitoring = true
	
	dynamic var name: String = ""
	dynamic var displayName: String = ""
	dynamic var installed: Bool = false
	dynamic var enabled: Bool = false
	dynamic var monitored: Bool = false
	dynamic var running: Bool = false
	
	override static func indexedProperties() -> [String] {
		return [
			"timestamp",
			"backgroundMonitoring",
			"name",
			"displayName",
			"installed",
			"enabled",
			"monitored",
			"running"
		]
	}
}
```
Code that triggers error:
```Swift
// This loop is just to print index/values of the Results to make sure the data is good
for (key, item) in data.enumerate() {
	print(key, (item as! Service).displayName)
}

let predicate = NSPredicate(format: "displayName BEGINSWITH[c] 'P'")
if let index = data.indexOf(predicate) { // This is the line that errors out
	return index
}
```
Console output of the For/In Loop .. indexOf should return 18
```
0 Apache SpamAssassin™
1 Apache Web Server
2 cPanel Daemon
3 cPanel DAV Daemon
4 cPanel DNS (Domain Name System) Admin Cache
5 cPanel Greylisting Daemon
6 cPanel Log and Bandwidth Processor
7 cPHulk Daemon
8 Cron Daemon
9 DNS (Domain Name System) Server
10 Exim Mail Server
11 Exim Mail Server (on another port)
12 FTP (File Transfer Protocol) Server
13 IMAP (Internet Mail Access Protocol) Server
14 IP (Internet Protocol) Aliases
15 Mailman
16 MySQL Server
17 Name Service Cache Daemon
18 Passive OS (Operating System) Fingerprinting Daemon
19 PHP-FPM service for cPanel Daemons
20 POP3 (Post Office Protocol 3) Server
21 rsyslog System Logger Daemon
22 SSH (Secure Shell) Daemon
23 TailWatch Daemon
24 TaskQueue Processor
```

indexOf goes to RealmSwift/Results.swift
```Swift
public func indexOf(predicate: NSPredicate) -> Int? {
        return notFoundToNil(rlmResults.indexOfObjectWithPredicate(predicate)) // Triggers here
 }
```
And finally in RealmSwift/Util.swift
```Swift
internal func notFoundToNil(index: UInt) -> Int? {
    if index == UInt(NSNotFound) {
        return nil
    }
    return Int(index) // Here
}
```
Stack trace
```
#0	0x000000010152c224 in notFoundToNil(UInt) -> Int? at /Users/REMOVED/xcode/iOS/REMOVED/Pods/RealmSwift/RealmSwift/Util.swift:28
#1	0x0000000101528634 in Results.indexOf<A where ...>(NSPredicate) -> Int? at /Users/REMOVED/xcode/iOS/REMOVED/Pods/RealmSwift/RealmSwift/Results.swift:143
#2	0x00000001001d3688 in TableViewDataSource.tableView<A where ...>(UITableView, sectionForSectionIndexTitle : String, atIndex : Int) -> Int at /Users/REMOVED/xcode/iOS/REMOVED/REMOVED/TableViewDataSource.swift:120
#3	0x00000001001d3850 in @objc TableViewDataSource.tableView<A where ...>(UITableView, sectionForSectionIndexTitle : String, atIndex : Int) -> Int ()
#4	0x0000000185fe8308 in -[UITableView _sectionIndexChangedToIndex:title:] ()
#5	0x0000000185eec20c in -[UITableView _sectionIndexChanged:] ()
#6	0x0000000185cdbe50 in -[UIApplication sendAction:to:from:forEvent:] ()
#7	0x0000000185cdbdcc in -[UIControl sendAction:to:forEvent:] ()
#8	0x0000000185cc3a88 in -[UIControl _sendActionsForEvents:withEvent:] ()
#9	0x0000000185eec138 in -[UITableViewIndex _selectSectionForTouch:withEvent:] ()
#10	0x0000000185eebf18 in -[UITableViewIndex beginTrackingWithTouch:withEvent:] ()
#11	0x0000000185ce4544 in -[UIControl touchesBegan:withEvent:] ()
#12	0x0000000185cdb168 in -[UIWindow _sendTouchesForEvent:] ()
#13	0x0000000185cd3e30 in -[UIWindow sendEvent:] ()
#14	0x0000000185ca44cc in -[UIApplication sendEvent:] ()
#15	0x0000000185ca2794 in _UIApplicationHandleEventQueue ()
#16	0x0000000180f68efc in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ ()
#17	0x0000000180f68990 in __CFRunLoopDoSources0 ()
#18	0x0000000180f66690 in __CFRunLoopRun ()
#19	0x0000000180e95680 in CFRunLoopRunSpecific ()
#20	0x00000001823a4088 in GSEventRunModal ()
#21	0x0000000185d0cd90 in UIApplicationMain ()
#22	0x0000000100051b3c in main at /Users/REMOVED/xcode/iOS/REMOVED/REMOVED/AppDelegate.swift:19
#23	0x0000000180a368b8 in start ()

```

## Version of Realm and Tooling

```
ProductName:	Mac OS X
ProductVersion:	10.11.3
BuildVersion:	15D21

/Applications/Xcode.app/Contents/Developer
Xcode 7.2.1
Build version 7C1002

pod is /usr/local/bin/pod
pod is /usr/local/bin/pod
0.39.0
Realm (0.98.4)
RealmSwift (0.98.4)

bash is /bin/bash
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin15)

carthage is /usr/local/bin/carthage
carthage is /usr/local/bin/carthage
0.9.2
(not in use here)

git is /usr/local/git/bin/git
git is /usr/bin/git
git is /usr/local/git/bin/git
git version 2.5.0
```

**Comment by jpsim** [comment]
I'm having trouble reproducing your issue. I'm wondering what type `data` is and where that's coming from. If it was a `Results<Service>` then you'd get a compiler warning with Swift 2.1.1 stating that "Forced cast of 'Service' to same type has no effect".

The following is my attempt at reproducing your issue, but it runs fine:

```swift
class Service: Object {
    dynamic var id = 0
    dynamic var displayName = ""

    override static func indexedProperties() -> [String] {
        return [
            "id",
            "displayName"
        ]
    }
}

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?

    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {

        // Delete Realm file to make this sample app deterministic
        do {
            try NSFileManager.defaultManager().removeItemAtPath(Realm.Configuration.defaultConfiguration.path!)
        } catch {}

        do {
            let realm = try Realm()

            do {
                try realm.write {
                    realm.create(Service.self, value: [0, "Apache SpamAssassin™"])
                    realm.create(Service.self, value: [1, "Apache Web Server"])
                    realm.create(Service.self, value: [2, "cPanel Daemon"])
                    realm.create(Service.self, value: [3, "cPanel DAV Daemon"])
                    realm.create(Service.self, value: [4, "cPanel DNS (Domain Name System) Admin Cache"])
                    realm.create(Service.self, value: [5, "cPanel Greylisting Daemon"])
                    realm.create(Service.self, value: [6, "cPanel Log and Bandwidth Processor"])
                    realm.create(Service.self, value: [7, "cPHulk Daemon"])
                    realm.create(Service.self, value: [8, "Cron Daemon"])
                    realm.create(Service.self, value: [9, "DNS (Domain Name System) Server"])
                    realm.create(Service.self, value: [10, "Exim Mail Server"])
                    realm.create(Service.self, value: [11, "Exim Mail Server (on another port)"])
                    realm.create(Service.self, value: [12, "FTP (File Transfer Protocol) Server"])
                    realm.create(Service.self, value: [13, "IMAP (Internet Mail Access Protocol) Server"])
                    realm.create(Service.self, value: [14, "IP (Internet Protocol) Aliases"])
                    realm.create(Service.self, value: [15, "Mailman"])
                    realm.create(Service.self, value: [16, "MySQL Server"])
                    realm.create(Service.self, value: [17, "Name Service Cache Daemon"])
                    realm.create(Service.self, value: [18, "Passive OS (Operating System) Fingerprinting Daemon"])
                    realm.create(Service.self, value: [19, "PHP-FPM service for cPanel Daemons"])
                    realm.create(Service.self, value: [20, "POP3 (Post Office Protocol 3) Server"])
                    realm.create(Service.self, value: [21, "rsyslog System Logger Daemon"])
                    realm.create(Service.self, value: [22, "SSH (Secure Shell) Daemon"])
                    realm.create(Service.self, value: [23, "TailWatch Daemon"])
                    realm.create(Service.self, value: [24, "TaskQueue Processor"])
                }
            } catch {
                print("error writing to Realm instance: \(error)")
            }

            let data = realm.objects(Service)

            // This loop is just to print index/values of the Results to make sure the data is good
            for (key, item) in data.enumerate() {
                print(key, item.displayName)
            }

            let predicate = NSPredicate(format: "displayName BEGINSWITH[c] 'P'")
            if let index = data.indexOf(predicate) {
                print(index) // prints 18, as expected
            }
        } catch {
            print("error creating Realm instance: \(error)")
        }
        return true
    }
}
```

**Comment by kbduvall** [comment]
`data` is a generic property, so I have to cast it to Service to access the object specific properties. The for-in loop was only just added to help in debugging this particular issue to make sure `data` contained the expected objects. Here's the class in it's original form for context. It appeared to be working fine on Realm 0.98.2. I First noticed the issue after updating to 0.98.4.

```Swift
typealias ItemToCellMappingBlock = (cell: CustomCellBase, item: Object) -> Void
class TableViewDataSource<O: Object>: NSObject, UITableViewDataSource {
	var data: Results<O>
	var sortKey: String? {
		didSet {
			if sortKey != nil {
				data = data.sorted(sortKey!, ascending: sortAsc)
			}
		}
	}
	var sortAsc: Bool = true {
		didSet {
			if sortKey != nil {
				data = data.sorted(sortKey!, ascending: sortAsc)
			}
		}
	}
	var identifier: String
	var mappingBlock: ItemToCellMappingBlock?
	
	var useIndex = false
	var indexKey: String?

	init(data: Results<O>, reuseIdentifier identifier: String, mappingBlock: ItemToCellMappingBlock) {
		self.data = data
		self.identifier = identifier
		self.mappingBlock = mappingBlock
	}
	func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return (useIndex && data.count > 0) ? 1 : data.count
	}
	
	func numberOfSectionsInTableView(tableView: UITableView) -> Int {
		return (useIndex) ? data.count : 1
	}

	func sectionIndexTitlesForTableView(tableView: UITableView) -> [String]? {
		if !useIndex || indexKey == nil || tableView.style == .Grouped { return nil }
		
		var output: [String] = []
		var orig: [Character] = []
		
		orig = data
			.filter({
				return ($0.valueForKey(indexKey!) as? String != nil)
					&& ($0.valueForKey(indexKey!) as! String != "")
			})
			.map({
				return ($0.valueForKey(indexKey!) as! String).uppercaseString.characters.first!
			})
		
		let unique = Array(Set(orig)).sort({$0 < $1})
		
		for char in Globals.alphabet.characters {
			if let _ = unique.indexOf(char) {
				output.append(String(char))
			}
			output.append(" ")
		}
		return output
	}
	
	func tableView(tableView: UITableView, sectionForSectionIndexTitle title: String, atIndex index: Int) -> Int {
		if !useIndex || indexKey == nil || tableView.style == .Grouped { return 0 }
		var t = title
		if title == " " {
			if index == 0 { return 0 }
			t = sectionIndexTitlesForTableView(tableView)![index - 1]
		}
		
		let predicate = NSPredicate(format: "%K BEGINSWITH[c] %@", indexKey!, t)
		if let index = data.indexOf(predicate) {
			return index
		}
		return data.count - 1
	}
	
	func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
		return tableView.delegate?.canEditRow(tableView, indexPath: indexPath) ?? false
	}
	
	func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
		let item = data[(useIndex) ? indexPath.section : indexPath.row]
		if let cell = tableView.dequeueReusableCellWithIdentifier(identifier, forIndexPath: indexPath) as? CustomCellBase, mapper = mappingBlock {
			cell.indexPath = indexPath
			mapper(cell: cell, item: item)
			return cell
		}
		let cell = CustomCellBase(style: UITableViewCellStyle.Value2, reuseIdentifier: identifier)
		cell.textLabel?.text = "Unable To Load Cell Data"
		return cell
	}
}
```

And in a controller: (`services` is a `List<Service>` list property in a `Server` object, which is contained in an `Account` object)
```Swift
services = account?.server?.services.filter("installed = true").sorted("displayName")
tableData = TableViewDataSource<Service>(data: services!, reuseIdentifier: "CellWithBadge") { cell, item in
	self.configureCell(cell as! CellWithBadge, item: item as! Service)
}
tableData.useIndex = true
tableData.indexKey = "displayName"

tableView.dataSource = tableData

/// Other table initializing code
```

The table index appears as it should, an A-Z character index for available service display names, showing only the letters for services listed in the table, and blank spaces for letters that don't have a corresponding service name. But when the index is tapped, that's when it crashes out.

**Comment by jpsim** [comment]
I've copy pasted your code into a sample project in an effort to reproduce your issue here: https://github.com/jpsim/GH3319

I've been unable to trigger the crash you've reported so far. I'd really appreciate if you could make a PR towards that repo with whatever changes are necessary to reproduce the crash. Please don't post any more partial code snippets here because it's very time consuming to fill in the gaps, especially when they contain code that's required to reproduce the issue.

**Comment by kbduvall** [comment]
After more investigation I noticed that it was only crashing on one of the two accounts set up. 

Both accounts have identical values for the displayName properties in their respective Service objects. 

Comparing the data from each account didn't show anything that would hint at being a cause of the crash as the method was comparing identical lists of service names.

So, instead of using Realm's indexOf, I changed it to use Swifts built in indexOf:
```Swift
if let index = data.indexOf({ (($0.valueForKey(indexKey) as? String) ?? "").uppercaseString.hasPrefix(t) }) {
	return index
}
```
And now it's working as expected.

I don't know why Realm is having a problem performing the search while Swift's default implementation isn't. With both datasets being essentially the same, I can't pinpoint the offending data or code.

**Comment by jpsim** [comment]
> So, instead of using Realm's indexOf, I changed it to use Swifts built in indexOf:

I'd really discourage you from doing this, because it will create Swift objects for each item being searched over. Also, if there's a bug in Realm, we'd really appreciate if you could help identify it. Even if you find a workaround, others might hit this issue in the future too.

**Comment by kbduvall** [comment]
I was able to duplicate it in a single view test project. I tried making a pull request to GH3319, but it's telling me I don't have permission.

**Comment by kbduvall** [comment]
I uploaded it to a new repo: https://github.com/kbduvall/GH3319

**Comment by bdash** [comment]
Code inspection reveals that `-[RLMResults indexOfObjectWithPredicate:]` is failing to pass the result of `Results::index_of` through `RLMConvertNotFound`, which I think would result in a failure of this nature (`index_of` returns an unsigned value with the maximum value representing not found. This maximum value cannot be represented in a signed Swift `Int`. `RLMConvertNotFound` maps the value to `NSNotFound` which does fit in an `Int`, and is what clients of the Cocoa API would expect to receive). Another possible oversight is that other calls to `index_of` within `RLMResults` are also wrapped in a `translateErrors` block, but the call within `-indexOfObjectWithPredicate:` is not.

**Comment by bdash** [comment]
After further thought, there must be another issue in `-[RLMResults indexOfObjectWithPredicate:]` as it shouldn't be possible for `_results.index_of` to return `realm::not_found`. For some reason the call to `Query::find` is returning a row that's not included in the restricting link view.

**Comment by bdash** [comment]
I filed a core issue since `Query::find` does not appear to be behaving as documented in this case, and the change I mentioned above relating to `RLMConvertNotFound` would merely replace a crash with incorrect results. Note that `Query::find` is only used by `-indexOfObjectWithPredicate:`. Querying in general uses `Query::find_all` which does not have the problem that `Query::find` has.

**Comment by kbduvall** [comment]
Is there any activity on this?

**Comment by bdash** [comment]
We're still waiting on a fix to the core database engine issue that is at the heart of this. I'll take a look at whether we can work around the underlying issue in Realm Cocoa later today. I suspect that we will be able to by trading performance for correctness.

**Comment by kbduvall** [comment]
It's not a huge rush. It'll be a bit before I publish. I just want to git rid of my Swift workaround before I do for the performance reasons you stated above. Due to the nature of my app, every user will have a different dataset from a different server, with a wide range of API versions, so I can't reliably use Realms indexOf anywhere in the app.

**Comment by bdash** [comment]
We were able to find a workaround with very little performance impact.
